apps/passport-server/src/services/userService.ts

```diff
@@ -148,55 +148,66 @@ export class UserService {
-      } else if (
-        !(await this.emailTokenService.checkTokenCorrect(email, emailToken))
-      ) {
-        throw new Error(
-          `Wrong token. If you got more than one email, use the latest one.`
-        );
-      } else if (user.email !== email) {
-        throw new Error(`Email mismatch.`);
-      }
-
-      // Save commitment to DB.
-      logger(`[ZUID] Saving new commitment: ${commitment}`);
-      const uuid = await insertCommitment(dbPool, {
-        email,
-        commitment
-      });
+    const user = await fetchZuzaluUser(this.context.dbPool, email);
+
+    if (user == null) {
+      throw new PCDHTTPError(403, `${email} doesn't have a ticket`);
+    }
 
-      // Reload Merkle trees
-      await this.semaphoreService.reload();
-      const newUser = await this.semaphoreService.getUserByUUID(uuid);
-      if (newUser == null) {
-        throw new Error(`${uuid} not found`);
-      } else if (newUser.commitment !== commitment) {
-        throw new Error(`Commitment mismatch`);
-      }
-
-      // Return user, including UUID, back to Passport
-      const zuzaluUser = newUser as User;
-      const jsonP = JSON.stringify(zuzaluUser);
-      logger(`[ZUID] Added new Zuzalu user: ${jsonP}`);
-
-      res.json(zuzaluUser);
-    } catch (e: any) {
-      logger(e);
-      this.rollbarService?.reportError(e);
-      res.status(500).send(e.message);
+    if (!(await this.emailTokenService.checkTokenCorrect(email, emailToken))) {
+      throw new PCDHTTPError(
+        403,
+        `Wrong token. If you got more than one email, use the latest one.`
+      );
+    }
+
+    if (user.email !== email) {
+      throw new PCDHTTPError(403, `Email mismatch.`);
     }
+
+    // Save commitment to DB.
+    logger(`[ZUID] Saving new commitment: ${commitment}`);
+    const uuid = await insertCommitment(this.context.dbPool, {
+      email,
+      commitment
+    });
+
+    // Reload Merkle trees
+    await this.semaphoreService.reload();
+
+    const newUser = (await this.semaphoreService.getUserByUUID(
+      uuid
+    )) as LoggedInZuzaluUser;
+
+    if (newUser == null) {
+      throw new PCDHTTPError(404, `user with id '${uuid}' not found`);
+    }
+
+    if (newUser.commitment !== commitment) {
+      throw new PCDHTTPError(403, `commitment mismatch`);
+    }
+
+    logger(`[ZUID] Added new Zuzalu user`, newUser);
+    res.status(200).json(newUser satisfies ZupassUserJson);
   }
 
+  /**
+   * If the service is not ready, returns a 500 server error.
+   * If the user does not exist, returns a 404.
+   * Otherwise returns the user.
+   */
   public async handleGetZuzaluUser(uuid: string, res: Response): Promise<void> {
     logger(`[ZUID] Fetching user ${uuid}`);
-    const user = await this.semaphoreService.getUserByUUID(uuid);
-    if (!user) res.status(404);
-    res.json(user || null);
+
+    const user = (await this.semaphoreService.getUserByUUID(
+      uuid
+    )) as LoggedInZuzaluUser;
```

Comment:
If the UUID passed in is for a PCDPass user and not a Zuzalu user, this will be a PCPass user irrespective of the type cast.

Reply:
I only know this because I'm looking at making this change myself as part of porting Zupass functionality to PCDPass!